packages/roosterjs-content-model-core/lib/coreApi/addUndoSnapshot/addUndoSnapshot.ts

```diff
@@ -11,19 +17,54 @@ import type { AddUndoSnapshot, Snapshot } from 'roosterjs-content-model-types';
  * when undo/redo to this snapshot
  */
 export const addUndoSnapshot: AddUndoSnapshot = (core, canUndoByBackspace, entityStates) => {
-    const { lifecycle, physicalRoot, undo } = core;
+    const { lifecycle, physicalRoot, logicalRoot, undo } = core;
     let snapshot: Snapshot | null = null;
 
     if (!lifecycle.shadowEditFragment) {
         // Need to create snapshot selection before retrieve innerHTML since HTML can be changed during creating selection when normalize table
         const selection = createSnapshotSelection(core);
         const html = physicalRoot.innerHTML;
 
+        // give plugins the chance to share entity states to include in the snapshot if the logical root is in an entity
+        const entityWrapper = findClosestEntityWrapper(logicalRoot, core.domHelper);
```

Comment:
We should also check if physicalRoot and logicalRoot are the same first. If they are the same, no need to find entity since it must return null, so save some time

Reply:
So in general this can be optimized to:

```ts
if (phy != logical) {
  const entityWrapper = !entityStates && ...
  if (entityWrapper ) { 
     ... // trigger event
   }

  logicalRootPath = ...
}
```

For the case that two roots are the same (99.999% case), we can even skip logicalRootPath so no need to change it when restore.